crates/ty_python_semantic/src/types/constraints.rs

```diff
@@ -273,1231 +239,1147 @@ impl<'db> ConstraintSet<'db> {
+            seen_constraints.insert(constraint);
+        });
+        let mut to_visit: Vec<(_, _)> = (seen_constraints.iter().copied())
+            .tuple_combinations()
             .collect();
-        for other_negative in &other.negative {
-            let Some(mut other_negative) = other_negative.clip_to_positive(db, &positive) else {
+
+        // Repeatedly pop constraint pairs off of the visit queue, checking whether each pair can
+        // be simplified.
+        let mut simplified = Node::Interior(self);
+        while let Some((left_constraint, right_constraint)) = to_visit.pop() {
+            // If the constraints refer to different typevars, they trivially cannot be compared.
+            // TODO: We might need to consider when one constraint's upper or lower bound refers to
+            // the other constraint's typevar.
+            let typevar = left_constraint.typevar(db);
+            if typevar != right_constraint.typevar(db) {
                 continue;
+            }
+
+            // Containment: The range of one constraint might completely contain the range of the
+            // other. If so, there are several potential simplifications.
+            let larger_smaller = if left_constraint.contains(db, right_constraint) {
+                Some((left_constraint, right_constraint))
+            } else if right_constraint.contains(db, left_constraint) {
+                Some((right_constraint, left_constraint))
+            } else {
+                None
             };
-            std::mem::swap(&mut previous, &mut current);
-            let mut previous_negative = previous.iter();
-            for self_negative in previous_negative.by_ref() {
-                match self_negative.intersect_negative(db, &other_negative) {
-                    None => {
-                        // We couldn't simplify the new hole relative to this existing holes, so
-                        // add the existing hole to the result. Continue trying to simplify the new
-                        // hole against the later existing holes.
-                        current.push(self_negative.clone());
-                    }
+            if let Some((larger_constraint, smaller_constraint)) = larger_smaller {
+                // larger ∨ smaller = larger
+                simplified = simplified.substitute_union(
+                    db,
+                    larger_constraint.when_true(),
+                    smaller_constraint.when_true(),
+                    Node::new_satisfied_constraint(db, larger_constraint.when_true()),
+                );
+
+                // larger ∧ smaller = smaller
+                simplified = simplified.substitute_intersection(
+                    db,
+                    larger_constraint.when_false(),
+                    smaller_constraint.when_false(),
+                    Node::new_satisfied_constraint(db, larger_constraint.when_false()),
+                );
```

Comment:
I might be misunderstanding, but it seems like the attempted simplification here doesn't match the comment? The simplification I'd expect to match the comment would be this one:

```suggestion
                // larger ∧ smaller = smaller
                simplified = simplified.substitute_intersection(
                    db,
                    larger_constraint.when_true(),
                    smaller_constraint.when_true(),
                    Node::new_satisfied_constraint(db, smaller_constraint.when_true()),
                );
```

It seems like the currently implemented simplification is `¬smaller ∧ ¬larger = ¬larger`, which is also correct, but not the same thing.